Configuration validation failed: Could not resolve placeholder ${auto.test.resources.datasources.default.username}

### Expected Behavior

I get this error: 

> Configuration validation failed: Could not resolve placeholder ${auto.test.resources.datasources.default.username}

[micronaut-data-jdbc-repository-maven-java.zip](https://github.com/user-attachments/files/26049763/micronaut-data-jdbc-repository-maven-java.zip)

```
./mvnw test
WARNING: A restricted method in java.lang.System has been called
WARNING: java.lang.System::load has been called by org.fusesource.jansi.internal.JansiLoader in an unnamed module (file:/Users/sdelamo/.m2/wrapper/dists/apache-maven-3.9.13-bin/6afe068c/apache-maven-3.9.13/lib/jansi-2.4.2.jar)
WARNING: Use --enable-native-access=ALL-UNNAMED to avoid a warning for callers in this module
WARNING: Restricted methods will be blocked in a future release unless native access is enabled

[INFO] Scanning for projects...
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for example.micronaut:default:jar:0.1
[WARNING] 'dependencyManagement.dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: com.vladsch.flexmark:flexmark-html2md-converter:jar -> version ${html2md.converter.version} vs ${flexmark.version} @ io.micronaut.platform:micronaut-platform:5.0.0-SNAPSHOT, /Users/sdelamo/.m2/repository/io/micronaut/platform/micronaut-platform/5.0.0-SNAPSHOT/micronaut-platform-5.0.0-SNAPSHOT.pom, line 3092, column 19
[WARNING] 
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING] 
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING] 
[INFO] 
[INFO] ---------------------< example.micronaut:default >----------------------
[INFO] Building default 0.1
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- enforcer:3.6.2:enforce (micronaut-enforce) @ default ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.version.RequireJavaVersion passed
[INFO] Rule 1: io.micronaut.maven.enforcer.CheckSnakeYaml passed
[INFO] 
[INFO] --- mn:5.0.0-M1:generate-import-factory (default-generate-import-factory) @ default ---
[INFO] 
[INFO] --- mn:5.0.0-M1:generate-openapi-generic (default-generate-openapi-generic) @ default ---
[INFO] 
[INFO] --- mn:5.0.0-M1:generate-openapi-client (default-generate-openapi-client) @ default ---
[INFO] 
[INFO] --- mn:5.0.0-M1:generate-openapi-server (default-generate-openapi-server) @ default ---
[INFO] 
[INFO] --- mn:5.0.0-M1:generate-jsonschema (default-generate-jsonschema) @ default ---
[INFO] 
[INFO] --- resources:3.5.0:resources (default-resources) @ default ---
[INFO] Copying 3 resources from src/main/resources to target/classes
[INFO] 
[INFO] --- compiler:3.15.0:compile (default-compile) @ default ---
[INFO] Nothing to compile - all classes are up to date.
[INFO] 
[INFO] --- resources:3.5.0:testResources (default-testResources) @ default ---
[INFO] Copying 0 resource from src/test/resources to target/test-classes
[INFO] 
[INFO] --- compiler:3.15.0:testCompile (default-testCompile) @ default ---
[INFO] Nothing to compile - all classes are up to date.
[INFO] 
[INFO] --- mn:5.0.0-M1:start-testresources-service (default-start-testresources-service) @ default ---
[info] Starting Micronaut Test Resources service, version 4.0.0-M1
[test-resources-service] 11:10:29.977 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments: [test]
[test-resources-service] 11:10:29.985 [ForkJoinPool.commonPool-worker-11] INFO  i.m.t.e.TestResourcesResolverLoader - Loaded 2 test resources resolvers: io.micronaut.testresources.mysql.MySQLTestResourceProvider, io.micronaut.testresources.testcontainers.GenericTestContainerProvider
[test-resources-service] 11:10:30.011 [pool-1-thread-1] INFO  o.testcontainers.DockerClientFactory - Testcontainers version: 4.0.0-M1
[test-resources-service] 11:10:30.126 [pool-1-thread-1] INFO  o.t.d.DockerClientProviderStrategy - Loaded org.testcontainers.dockerclient.UnixSocketClientProviderStrategy from ~/.testcontainers.properties, will try it first
[test-resources-service] 11:10:30.131 [pool-1-thread-1] WARN  o.t.d.DockerClientProviderStrategy - DOCKER_HOST tcp://127.0.0.1:56364 is not listening
java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	at java.base/sun.nio.ch.Net.connect(Net.java:546)
	at java.base/sun.nio.ch.Net.connect(Net.java:535)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:585)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:284)
	at java.base/java.net.Socket.connect(Socket.java:659)
	at java.base/java.net.Socket.connect(Socket.java:597)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.lambda$test$3(DockerClientProviderStrategy.java:214)
	at org.testcontainers.shaded.org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.testcontainers.shaded.org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
	at org.testcontainers.shaded.org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at java.base/java.lang.Thread.run(Thread.java:1474)
[test-resources-service] 11:10:30.132 [pool-1-thread-1] WARN  o.t.d.DockerClientProviderStrategy - DOCKER_HOST tcp://127.0.0.1:56364 is not listening
java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	at java.base/sun.nio.ch.Net.connect(Net.java:546)
	at java.base/sun.nio.ch.Net.connect(Net.java:535)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:585)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:284)
	at java.base/java.net.Socket.connect(Socket.java:659)
	at java.base/java.net.Socket.connect(Socket.java:597)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.lambda$test$3(DockerClientProviderStrategy.java:214)
	at org.testcontainers.shaded.org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.testcontainers.shaded.org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
	at org.testcontainers.shaded.org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at java.base/java.lang.Thread.run(Thread.java:1474)
[test-resources-service] 11:10:30.285 [pool-1-thread-1] INFO  o.t.d.DockerClientProviderStrategy - Found Docker environment with local Unix socket (unix:///var/run/docker.sock)
[test-resources-service] 11:10:30.286 [pool-1-thread-1] INFO  o.testcontainers.DockerClientFactory - Docker host IP address is localhost
[test-resources-service] 11:10:30.299 [pool-1-thread-1] INFO  o.testcontainers.DockerClientFactory - Connected to docker: 
  Server Version: 29.2.1
  API Version: 1.53
  Operating System: Docker Desktop
  Total Memory: 48089 MB
  Labels: 
    com.docker.desktop.address=unix:///Users/sdelamo/Library/Containers/com.docker.docker/Data/docker-cli.sock
[test-resources-service] 11:10:30.302 [pool-1-thread-1] INFO  o.testcontainers.images.PullPolicy - Image pull policy will be performed by: DefaultPullPolicy()
[test-resources-service] 11:10:30.303 [pool-1-thread-1] INFO  o.t.utility.ImageNameSubstitutor - Image name substitution will be performed by: DefaultImageNameSubstitutor (composite of 'ConfigurationFileImageNameSubstitutor' and 'PrefixingImageNameSubstitutor')
[test-resources-service] 11:10:30.305 [pool-1-thread-1] INFO  o.testcontainers.DockerClientFactory - Checking the system...
[test-resources-service] 11:10:30.305 [pool-1-thread-1] INFO  o.testcontainers.DockerClientFactory - ✔︎ Docker server version should be at least 1.6.0
[test-resources-service] 11:10:30.432 [scheduled-executor-thread-1] INFO  i.m.t.server.ExpiryManager - Test resources server will automatically be shutdown if it doesn't receive requests for 60 minutes
[test-resources-service] 11:10:30.495 [main] INFO  i.m.t.server.TestResourcesService - A Micronaut Test Resources server is listening on port 65118, started in 648ms
[INFO] Test resources service already started on port 65118
[INFO] 
[INFO] --- mn:5.0.0-M1:validate-test-configuration (default-validate-test-configuration) @ default ---
[INFO] Validating Micronaut configuration (test)
[INFO] Established active environments: [test]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.938 s
[INFO] Finished at: 2026-03-17T11:10:31+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.micronaut.maven:micronaut-maven-plugin:5.0.0-M1:validate-test-configuration (default-validate-test-configuration) on project default: Configuration validation failed: Could not resolve placeholder ${auto.test.resources.datasources.default.username} -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

### Version

5.0.0-SNAPSHOT

--- a/json-schema-configuration-validator/src/main/java/io/micronaut/jsonschema/configuration/validator/AutoTestResourcesPropertyExpressionResolver.java
+++ b/json-schema-configuration-validator/src/main/java/io/micronaut/jsonschema/configuration/validator/AutoTestResourcesPropertyExpressionResolver.java
@@ -1,0 +1,59 @@
+/*
+ * Copyright 2017-2026 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.jsonschema.configuration.validator;
+
+import io.micronaut.context.env.PropertyExpressionResolver;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.value.PropertyResolver;
+
+import java.util.Optional;
+
+/**
+ * Provides placeholder fallback values for Micronaut Test Resources expressions.
+ */
+@Internal
+public final class AutoTestResourcesPropertyExpressionResolver implements PropertyExpressionResolver {
+    static final String PREFIX = "auto.test.resources.";
+    private static final String DUMMY_TEXT_VALUE = "test-resource";
+
+    @Override
+    public <T> Optional<T> resolve(
+        PropertyResolver propertyResolver,
+        ConversionService conversionService,
+        String expression,
+        Class<T> requiredType
+    ) {
+        if (!expression.startsWith(PREFIX)) {
+            return Optional.empty();
+        }
+        if (requiredType.isAssignableFrom(String.class)) {
+            return Optional.of(requiredType.cast(DUMMY_TEXT_VALUE));
+        }
+
+        Optional<T> converted = conversionService.convert(DUMMY_TEXT_VALUE, requiredType);
+        if (converted.isPresent()) {
+            return converted;
+        }
+
+        converted = conversionService.convert("0", requiredType);
+        if (converted.isPresent()) {
+            return converted;
+        }
+
+        return conversionService.convert("false", requiredType);
+    }
+}

--- a/json-schema-configuration-validator/src/main/resources/META-INF/services/io.micronaut.context.env.PropertyExpressionResolver
+++ b/json-schema-configuration-validator/src/main/resources/META-INF/services/io.micronaut.context.env.PropertyExpressionResolver
@@ -1,0 +1,1 @@
+io.micronaut.jsonschema.configuration.validator.AutoTestResourcesPropertyExpressionResolver

--- a/json-schema-configuration-validator/src/test/java/io/micronaut/jsonschema/configuration/validator/AutoTestResourcesPropertyExpressionResolverTest.java
+++ b/json-schema-configuration-validator/src/test/java/io/micronaut/jsonschema/configuration/validator/AutoTestResourcesPropertyExpressionResolverTest.java
@@ -1,0 +1,104 @@
+/*
+ * Copyright 2017-2026 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.jsonschema.configuration.validator;
+
+import io.micronaut.context.env.Environment;
+import io.micronaut.core.convert.MutableConversionService;
+import org.junit.jupiter.api.Test;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class AutoTestResourcesPropertyExpressionResolverTest {
+
+    private final AutoTestResourcesPropertyExpressionResolver resolver = new AutoTestResourcesPropertyExpressionResolver();
+
+    @Test
+    void nonAutoTestResourcesExpressionsRemainUnresolved() {
+        try (Environment environment = newEnvironment()) {
+            assertTrue(resolver.resolve(environment, MutableConversionService.create(), "datasources.default.username", String.class).isEmpty());
+        }
+    }
+
+    @Test
+    void autoTestResourcesExpressionsResolveToDummyString() {
+        try (Environment environment = newEnvironment()) {
+            assertEquals("test-resource", resolver.resolve(environment, MutableConversionService.create(), "auto.test.resources.datasources.default.username", String.class).orElseThrow());
+        }
+    }
+
+    @Test
+    void resolverUsesPrimaryConversionResultWhenAvailable() {
+        try (Environment environment = newEnvironment()) {
+            URI uri = resolver.resolve(environment, MutableConversionService.create(), "auto.test.resources.datasources.default.url", URI.class).orElseThrow();
+            assertEquals(URI.create("test-resource"), uri);
+        }
+    }
+
+    @Test
+    void resolverFallsBackToFalseLiteralWhenRequired() {
+        try (Environment environment = newEnvironment()) {
+            MutableConversionService conversionService = MutableConversionService.create();
+            conversionService.addConverter(String.class, FalseOnlyValue.class, value -> "false".equals(value) ? new FalseOnlyValue(value) : null);
+
+            FalseOnlyValue value = resolver.resolve(environment, conversionService, "auto.test.resources.flag", FalseOnlyValue.class).orElseThrow();
+            assertEquals("false", value.value());
+            assertFalse("0".equals(value.value()));
+        }
+    }
+
+    @Test
+    void resolverFallsBackToZeroLiteralWhenRequired() {
+        try (Environment environment = newEnvironment()) {
+            MutableConversionService conversionService = MutableConversionService.create();
+            conversionService.addConverter(String.class, ZeroOnlyValue.class, value -> "0".equals(value) ? new ZeroOnlyValue(value) : null);
+
+            ZeroOnlyValue value = resolver.resolve(environment, conversionService, "auto.test.resources.port", ZeroOnlyValue.class).orElseThrow();
+            assertEquals("0", value.value());
+        }
+    }
+
+    record FalseOnlyValue(String value) {
+    }
+
+    record ZeroOnlyValue(String value) {
+    }
+
+    private static Environment newEnvironment() {
+        ClassLoader classLoader = AutoTestResourcesPropertyExpressionResolverTest.class.getClassLoader();
+        return Environment.create(new io.micronaut.context.ApplicationContextConfiguration() {
+            @Override
+            public Optional<Boolean> getDeduceEnvironments() {
+                return Optional.of(false);
+            }
+
+            @Override
+            public List<String> getEnvironments() {
+                return List.of();
+            }
+
+            @Override
+            public ClassLoader getClassLoader() {
+                return classLoader;
+            }
+        }).start();
+    }
+}

--- a/json-schema-configuration-validator/src/test/java/io/micronaut/jsonschema/configuration/validator/cli/ConfigurationJsonSchemaValidatorCliTest.java
+++ b/json-schema-configuration-validator/src/test/java/io/micronaut/jsonschema/configuration/validator/cli/ConfigurationJsonSchemaValidatorCliTest.java
@@ -752,6 +752,34 @@ class ConfigurationJsonSchemaValidatorCliTest {
     }
 
     @Test
+    void cliResolvesAutoTestResourcesPlaceholdersDuringConfigurationValidation() throws Exception {
+        Path cp = tempDir.resolve("cp-auto-test-resources");
+        Files.createDirectories(cp);
+        Files.writeString(cp.resolve("application.properties"), String.join("\n",
+            "micronaut.application.name=${auto.test.resources.micronaut.application.name}",
+            ""
+        ), StandardCharsets.UTF_8);
+
+        Path out = tempDir.resolve("out-auto-test-resources");
+        ByteArrayOutputStream errCapture = new ByteArrayOutputStream();
+        PrintStream err = new PrintStream(errCapture, true, StandardCharsets.UTF_8);
+
+        String classpath = cp + File.pathSeparator + System.getProperty("java.class.path");
+        int exit = ConfigurationJsonSchemaValidatorCli.run(new String[] {
+            "--classpath", classpath,
+            "--environments", "test",
+            "--out", out.toString(),
+            "--format", "json"
+        }, System.out, err);
+
+        assertEquals(0, exit, () -> "Unexpected stderr:\n" + errCapture.toString(StandardCharsets.UTF_8));
+        assertTrue(Files.exists(out.resolve("configuration-errors.json")));
+
+        String stderr = errCapture.toString(StandardCharsets.UTF_8);
+        assertFalse(stderr.contains("Could not resolve placeholder"), () -> "Unexpected stderr:\n" + stderr);
+    }
+
+    @Test
     void cliRendersRelativeOriginPathsWhenConfigured() throws Exception {
         Path project = tempDir.resolve("project");
         Path resources = project.resolve("src/main/resources");

